src/main/java/org/opensearch/knn/index/KNNVectorDVLeafFieldData.java

```diff
@@ -48,6 +50,44 @@ public ScriptDocValues<float[]> getScriptValues() {
 
     @Override
     public SortedBinaryDocValues getBytesValues() {
-        throw new UnsupportedOperationException("knn vector field '" + fieldName + "' doesn't support sorting");
+        try {
+            final BinaryDocValues binaryDocValues = DocValues.getBinary(reader, fieldName);
+            SortedBinaryDocValues sortedBinaryDocValues = new SortedBinaryDocValues() {
+
+                private boolean docExists = false;
+                float[] floats = null;
+                int pos = 0;
+                BytesRefBuilder bytesRefBuilder = new BytesRefBuilder();
+
+                @Override
+                public boolean advanceExact(int doc) throws IOException {
+                    if (binaryDocValues.advanceExact(doc)) {
+                        docExists = true;
+                        floats = vectorDataType.getVectorFromDocValues(binaryDocValues.binaryValue());
+                        pos = 0;
+                        return docExists;
+                    }
+                    docExists = false;
+                    return docExists;
+                }
+
+                @Override
+                public int docValueCount() {
+                    return docExists ? floats.length : 0;
```

Comment:
This is incorrect right? Looking at comment:
```
Retrieves the number of values for the current document. This must always be greater than zero. It is illegal to call this method after advanceExact(int) returned false.
```

I think it should be 1 because we dont support arrays of vectors. For this, it assumes that we can get values per float, which seems like a potential issue as the data type is actually float[].

Reply:
yes correctly!

Reply:
> Im not sure I understand. The docValueCount should be 1 if it exists. 0 otherwise. I believe doc value count here refers to multiple values per field. This turns into kind of an array type:

@jmazanec15 we use `BinaryDocValues` to store vector data as doc_values, so the vector count is 1. in `KNNVectorDVLeafFieldData#getBytesValues` I want to make `docvalue_fields` search return a json content which is `utf-8` characters not only the binary format data. so I use `docValueCount` stands for array length and nextValue for number chars.  and this can return a array format data when return a docvalue_fields.

Reply:
Ah I see. But then we are breaking the symmetry where `public ScriptDocValues<float[]> getScriptValues() {` returns representation of float[]:
```
    public float[] getValue() {
        if (!docExists) {
            String errorMessage = String.format(
                "One of the document doesn't have a value for field '%s'. "
                    + "This can be avoided by checking if a document has a value for the field or not "
                    + "by doc['%s'].size() == 0 ? 0 : {your script}",
                fieldName,
                fieldName
            );
            throw new IllegalStateException(errorMessage);
        }
        try {
            return vectorDataType.getVectorFromDocValues(binaryDocValues.binaryValue());
        } catch (IOException e) {
            throw ExceptionsHelper.convertToOpenSearchException(e);
        }
    }
```

Why cant we have `public BytesRef nextValue() throws IOException {` return the whole string represetation of the current vector?

Reply:
> @jmazanec15 as I see, we say synthetic source field is type XContentType.JSON.xContent . but SortedBinaryDocValues#BytesRef nexValue() is bin bytes array. I am not sure we need trans the bin bytes in DocValuesFormat to utf-8 JSON like format bytes string.
> 
> also, I see we are trying to reconstruct the vector format with KnnVectorsFormat. so I think in KnnVectorsFormat we can simply rewrite docvalues with JSON.xContent format.
> 
> but I am not sure Which SortedBinaryDocValues#BytesRef nexValue() format is better(bin, or one double value).
> 
> what do you think which is better.

@luyuncheng lets continue this discussion here to preserve context of the discussion. I think that per field, the vector is the binary doc value. For instance, in https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/VectorField.java#L19-L21, we set the bytes value as `this.setBytesValue(floatToByte);`. Can we follow this same logic for sorted binary doc values here? Then, the consumer can do what they need to do to convert the bytes to the proper type. My worry is, if we do it per-dimension, then it will be messy if we ever do support arrays of knn_vectors. 

@navneet1v what do you think here?

Reply:
I read the discussion, I am aligned with what @jmazanec15 is recommending here. Lets just reverse the logic what were are doing for converting vectors to binary doc values.

The idea of having multiple values in docvalues is more for multi-valued field. But for KNN vector field the vector is a single value for us.

Reply:
> The idea of having multiple values in docvalues is more for multi-valued field. But for KNN vector field the vector is a single value for us.

@jmazanec15 @navneet1v   I think it is difficult to use one `bytesRef` for one KNN vector. because we are trying docValues. and in OpenSearch `DocValueFetcher#fetchValues` :

```
    @Override
    public List<Object> fetchValues(SourceLookup lookup) throws IOException {
        if (false == leaf.advanceExact(lookup.docId())) {
            return emptyList();
        }
        List<Object> result = new ArrayList<Object>(leaf.docValueCount());
        for (int i = 0, count = leaf.docValueCount(); i < count; ++i) {
            result.add(leaf.nextValue());
        }
        return result;
    }
```

it use a list object to add all `doc_values`. so when we use `one docCount` with bytesRef or rewrite Leaf return floats it would cause doc_values field response like following:

```
{"vector_field":["[1.5,2.5]"]}  result list is one XContent array object
{"vector_field":["1.5 2.5"]} result list is one XContent value object
{"vector_field1":[[1.5,2.5]]} result list is one array object
```

these response can not make vector_field in reindex or other scenarios. so at 1st version I prefer to use docValueCount as floats.length, and I will create a new pr for OpenSearch which can support `DocValueFetcher#fetchValues` return an object not a list<Object>.

Reply:
in latest commits, I use `DocValueFetcher.Leaf` to return `knn docvalues`. because `SortedBinaryDocValues` return a bytesRef in result list, and result would be like a string array

```
{"vector_field1":["1.5","2.5"]} result list is SortedBinaryDocValues and count is floats.length
```

in following code results like:

https://github.com/opensearch-project/k-NN/blob/be60d425a12658be510e3e021da84415b136aa0f/src/main/java/org/opensearch/knn/index/KNNVectorDVLeafFieldData.java#L66-L92

```
{"vector_field1":[1.5,2.5]} result list is DocValueFetcher.Leaf return floats with count is floats.length
```

Reply:
If we accept doc response like:

```
 {"fields":{"vector_field1":[[1.5,2.5]]}}
```

we can process the response at `KNNFetchSubPhase` with 1st element and rebuild into `_source` field

Reply:
I see. So right now, we do not accept an array of vectors for a field. So I think that `f.fetcher.fetchValues(hitContext.sourceLookup())` will always return a list of size 1. With this assumption, can we then do something like:
```
// (not including type checks/validation for sake of brevity)
maps.put(f.field, f.fetcher.fetchValues(hitContext.sourceLookup()).get(0));
```

Which would then lead to:
```
 {"fields":{"vector_field1":[1.5,2.5]}}
```

Reply:
> I see. So right now, we do not accept an array of vectors for a field. So I think that f.fetcher.fetchValues(hitContext.sourceLookup()) will always return a list of size 1. With this assumption, can we then do something like:

@jmazanec15  LGTM 

update at https://github.com/opensearch-project/k-NN/pull/1571/commits/cf98300eebd0b1e0a25a2110de73077c101a24ff

Reply:
+1 on Jacks comment. I tested this with other fields example.

```
PUT table
{
  "settings": {
    "index": {
      "knn": true
    }
  }, 
  "mappings": {
    "properties": {
      "my_vector1": {
        "type": "knn_vector",
				"dimension": 1,
				"method": {
					"name": "hnsw",
					"space_type": "l2",
					"engine": "faiss"
				}
      },
      "rowId": {
        "type": "keyword",
        "store": true
      },
      "applicationId": {
        "type": "keyword",
        "store": true
      }
    }
  }
}




PUT /_bulk?refresh=true
{ "index": { "_index": "table" } }
{"my_vector1":[1], "rowId": "recWtYS3a5LENkhte", "applicationId": "appw3EJpjHiON9qhd", "id": "1"}
{ "index": { "_index": "table"} }
{"my_vector1":[2], "rowId": "recZRfK7DnN0K0ZKM", "applicationId": "appw3EJpjHiON9qhd", "id": "2"}
{ "index": { "_index": "table"} }
{"my_vector1":[3], "rowId": "ABCrecsCBwqA7E7OWvqw", "applicationId": "appw3EJpjHiON9qhd", "id": "3"}
{ "index": { "_index": "table"} }
{"my_vector1":[4], "rowId": "recWtYS3a5LENkhte", "applicationId": "appw5EJpjHiON9qhd", "id": "4"}
{ "index": { "_index": "table"} }
{"my_vector1":[5], "rowId": "recZRfK7DnN0K0ZKM", "applicationId": "appw6EJpjHiON9qhd", "id": "5"}
{ "index": { "_index": "table"} }
{"my_vector1":[6], "rowId": "recsCBwqA7E7OWvqw", "applicationId": "appw3EJpjHiON9qhd", "id": "6"}

POST table/_search
{
  "stored_fields": ["applicationId"]
}
```

#### Response
```

{
  "took": 1,
  "timed_out": false,
  "_shards": {
    "total": 5,
    "successful": 5,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 6,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": "table",
        "_id": "nUPD8o4BA3Dhnn4-ORLg",
        "_score": 1,
        "fields": {
          "applicationId": [
            "appw3EJpjHiON9qhd"
          ]
        }
      },
      {
        "_index": "table",
        "_id": "oEPD8o4BA3Dhnn4-ORLg",
        "_score": 1,
        "fields": {
          "applicationId": [
            "appw5EJpjHiON9qhd"
          ]
        }
      },
      {
        "_index": "table",
        "_id": "okPD8o4BA3Dhnn4-ORLg",
        "_score": 1,
        "fields": {
          "applicationId": [
            "appw3EJpjHiON9qhd"
          ]
        }
      },
      {
        "_index": "table",
        "_id": "nkPD8o4BA3Dhnn4-ORLg",
        "_score": 1,
        "fields": {
          "applicationId": [
            "appw3EJpjHiON9qhd"
          ]
        }
      },
      {
        "_index": "table",
        "_id": "n0PD8o4BA3Dhnn4-ORLg",
        "_score": 1,
        "fields": {
          "applicationId": [
            "appw3EJpjHiON9qhd"
          ]
        }
      },
      {
        "_index": "table",
        "_id": "oUPD8o4BA3Dhnn4-ORLg",
        "_score": 1,
        "fields": {
          "applicationId": [
            "appw6EJpjHiON9qhd"
          ]
        }
      }
    ]
  }
}
```

Reply:
> That being said, users would expect [[1.5,2.5]] from the response for fields. If we return [1.5,2.5], this would break user expectations.

@jmazanec15  in https://github.com/opensearch-project/k-NN/commit/cf98300eebd0b1e0a25a2110de73077c101a24ff  when we using doc_values like

```
POST some_index/_search
{
  "docvalue_fields": [
    "vector_field1",
    "vector_field2",
  ],
  "_source": false
}'
```

it would return `{"fields":{"vector_field1":[[1.5,2.5]]}}` as you thinks. and in `FetchSubPhase` i only use the 1st vector value for source field. i think it is ok for the scenarios as you said

Reply:
because `DocValueFetcher#fetchValues` do the same thing as your said: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/mapper/SourceValueFetcher.java#L81-L82